pallets/api/src/messaging/mod.rs

```diff
@@ -234,151 +249,174 @@ pub mod pallet {
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
-		#[pallet::call_index(0)]
-		#[pallet::weight(Weight::zero())] // todo: benchmarking after consolidating storage
-		pub fn send(_origin: OriginFor<T>, _id: MessageId) -> DispatchResult {
-			// e.g. Message::StateQuery { dest: Parachain(1000), storage_keys: vec![] }
-			// e.g. Message::Transact { dest: Parachain(1000), call: vec![] }
-			todo!("Reserved for messaging abstractions")
-		}
-
 		// TODO: does ismp allow querying to ensure that specified para id is supported?
 		#[pallet::call_index(1)]
-		#[pallet::weight(Weight::zero() // todo: benchmarking after consolidating storage
-			// Add any additional gas limit specified for callback execution
-			.saturating_add(callback.map(|cb| {
-				T::CallbackExecutor::execution_weight().saturating_add(cb.weight)
-			}).unwrap_or_default())
-		)]
+		#[pallet::weight(Weight::zero())]
 		pub fn ismp_get(
 			origin: OriginFor<T>,
 			id: MessageId,
 			message: ismp::Get<T>,
 			fee: BalanceOf<T>,
-			callback: Option<Callback>,
+			callback: Option<Callback<T::AccountId>>,
 		) -> DispatchResult {
 			let origin = ensure_signed(origin)?;
+			ensure!(!Messages::<T>::contains_key(&origin, &id), Error::<T>::MessageExists);
+
+			let deposit = calculate_protocol_deposit::<T, T::OnChainByteFee>(
+				ProtocolStorageDeposit::IsmpRequests,
+			) + calculate_message_deposit::<T, T::OnChainByteFee>() +
+				calculate_deposit_of::<T, T::OffChainByteFee, ismp::Get<T>>();
 
-			// Calculate deposit and place on hold.
-			let deposit = Self::calculate_deposit(
-				message.calculate_deposit() +
-					// IsmpRequests
-					(KeyLenOf::<IsmpRequests<T>>::get() as usize +
-						AccountIdOf::<T>::max_encoded_len() +
-						MessageId::max_encoded_len())
-					.saturated_into::<BalanceOf<T>>() *
-						T::ByteFee::get(),
-			);
 			T::Deposit::hold(&HoldReason::Messaging.into(), &origin, deposit)?;
 
 			// Process message by dispatching request via ISMP.
-			ensure!(!Messages::<T>::contains_key(&origin, &id), Error::<T>::MessageExists);
-			let commitment = T::IsmpDispatcher::default()
-				.dispatch_request(message.into(), FeeMetadata { payer: origin.clone(), fee })
-				.map_err(|_| Error::<T>::DispatchFailed)?;
-
-			// Store commitment for lookup on response, message for querying, response/timeout
-			// handling.
-			IsmpRequests::<T>::insert(&commitment, (&origin, id));
-			Messages::<T>::insert(&origin, id, Message::Ismp { commitment, callback, deposit });
-			Pallet::<T>::deposit_event(Event::<T>::IsmpGetDispatched {
-				origin,
-				id,
-				commitment,
-				callback,
-			});
-			Ok(())
+			let maybe_commitment = T::IsmpDispatcher::default()
+				.dispatch_request(message.into(), FeeMetadata { payer: origin.clone(), fee });
+
+			match maybe_commitment {
```

Comment:
hmm, this is an interesting addition.

If the `dispatch_request` fails, why aren't we just returning a dispatch error and rolling back the extrinsic? The contract will be able to decode the error directly and then decide what to do in the same block (instead of having to do another call). All dispatchables are transactional.

Reply:
Huh! transactional was opt in when i was working at Imbue, looks like this has changed to opt out!
Awesome.